tests/test_player.py

```diff
@@ -0,0 +1,17 @@
+from unittest import TestCase
+
+from card import Card
+from player import Player
+
+
+class PlayerTestCase(TestCase):
+    def test_all_possible_two_number_card_hand_combinations(self):
```

Comment:
> Am I violating principals by testing multiple hand combinations like this?

Not necessarily. If it's incredibly important for an app that all the combinations be tested, then doing something like this would be appropriate. However, This is very likely overkill. And it's not going to scale very well. In addition to that, it doesn't read very well. That being said, I'm not surprised you started with something like this. It's not uncommon for an overkill solution like this to be the first stab at testing. 

What would be more helpful would be having tests that show specific hands work. And those specific hands should fall in one of a couple categories:

* Happy path (something like 3, 7 = 10). 
* Edge cases (dupes: 6, 6 = 12, max: k, q = 20, min: 2, 3 = 4)
* Any combination that has existed as a bug in the past.

Were it me, I would make a helper method on `PlayerTestCase` called `self.assert_hand_score(self, cards, expected_count):`

Then, you could have a test like:

```python
def test_simple(self):
    self.assert_hand_score([3, 7], 10)
    self.assert_hand_score([6, 6], 12)
    self.assert_hand_score([2, 3, 4], 9)

def test_faces(self):
    ...

def test_aces(self):
    ...
```

From here, the body of the tests is incredibly easy to read. Adding a new combination is a matter of copying/pasting a line and updating the list we pass in. And we rely on the the helper method to create the player, set their hand, get the score, and do the assert. If we need to, we can look at the helper method to help figure out exactly what is going on. But the actual test is super easy to read and understand what exactly we want to accomplish.

In tests, it's better to be explicit over implicit. We don't want other devs or future us coming back and being like "what is going on here?". We can't always do that, but very often, we can.

Make sense?

Reply:
I see. I need to give more priority to readability.

Reply:
When you are working on your own, and no one is looking at your code, it's super easy to write things that are really hard to read. You get really close to the problem and it makes perfect sense to you since you are in the zone. Then six months later you look at it and you're all like" Oh God, what I have unleashed in this world. Please forgive me". At least that's what always happened to me. Still happens to me sometimes. It's not an accident that I was super excited when we finally hired a contractor. Not because I was excited because they would be there to write some of the code in my backlog (although I was happy about this), but because code review is that important.